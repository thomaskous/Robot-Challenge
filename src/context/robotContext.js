import { createContext } from 'react'

export const RobotContext = createContext();