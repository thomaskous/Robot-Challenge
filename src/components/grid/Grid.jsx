import './index.css'
import React, { useContext } from 'react'
import Cell from '../cell/Cell'
import Robot from '../robot/robot';
import { RobotContext } from '../../context/robotContext'

/* The Grid Component renders the cells as a 5x5 raster 
   and passes the Robot to the "active" cell. To do that it 
   has to consume the position and the facing from the RobotContext*/
function Grid({ rows, columns }) {
  const { position, facing } = useContext(RobotContext);

  const cells = [];
  let cellId = null;

  if (null !== position.x && null !== position.y && facing) {
    cellId = position.y + '' + position.x;
  }

  // use a for loop to create rows
  for (let i = rows; i > 0; i--) {
    // use a nested loop to create columns inside the rows
    for (let y = 0; y < columns; y++) {
      cells.push(
        <Cell key={i + '/' + y} id={`${y}${i - 1}`} data-testid='cell'>
          {(cellId === `${i - 1}${y}`) && <Robot facing={facing} />}
        </Cell>
      );
    }
  }

  return (
    <div className="grid">
      {cells}
    </div>
  )
}

export default Grid;