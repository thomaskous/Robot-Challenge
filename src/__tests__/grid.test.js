import React from 'react'
import { render } from '@testing-library/react'
import Grid from '../components/grid/Grid'
import { RobotContext } from '../context/robotContext';

describe('<Grid />', () => {
  it('renders an 3x3 <Grid /> with no robot', () => {
    const { container, getAllByTestId, queryByAltText } = render(
      <RobotContext.Provider value={{ position: { x: null, y: null }, facing: '' }}>
        <Grid rows={3} columns={3} />
      </RobotContext.Provider>
    )

    expect(getAllByTestId('cell').length).toEqual(9);
    expect(queryByAltText('robot')).toBeFalsy();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an 3x3 <Grid /> with a robot', () => {
    const { container, getAllByTestId, getByAltText } = render(
      <RobotContext.Provider value={{ position: { x: 1, y: 1 }, facing: 'N' }}>
        <Grid rows={3} columns={3} />
      </RobotContext.Provider>
    )

    expect(getAllByTestId('cell').length).toEqual(9);
    expect(getByAltText('robot')).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });
});