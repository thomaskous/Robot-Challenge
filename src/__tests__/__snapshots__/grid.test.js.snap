// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> renders an 3x3 <Grid /> with a robot 1`] = `
<div
  class="grid"
>
  <div
    class="cell"
    data-testid="cell"
    id="02"
  />
  <div
    class="cell"
    data-testid="cell"
    id="12"
  />
  <div
    class="cell"
    data-testid="cell"
    id="22"
  />
  <div
    class="cell"
    data-testid="cell"
    id="01"
  />
  <div
    class="cell"
    data-testid="cell"
    id="11"
  >
    <img
      alt="robot"
      class="robot N"
      src="/robot.png"
    />
  </div>
  <div
    class="cell"
    data-testid="cell"
    id="21"
  />
  <div
    class="cell"
    data-testid="cell"
    id="00"
  />
  <div
    class="cell"
    data-testid="cell"
    id="10"
  />
  <div
    class="cell"
    data-testid="cell"
    id="20"
  />
</div>
`;

exports[`<Grid /> renders an 3x3 <Grid /> with no robot 1`] = `
<div
  class="grid"
>
  <div
    class="cell"
    data-testid="cell"
    id="02"
  />
  <div
    class="cell"
    data-testid="cell"
    id="12"
  />
  <div
    class="cell"
    data-testid="cell"
    id="22"
  />
  <div
    class="cell"
    data-testid="cell"
    id="01"
  />
  <div
    class="cell"
    data-testid="cell"
    id="11"
  />
  <div
    class="cell"
    data-testid="cell"
    id="21"
  />
  <div
    class="cell"
    data-testid="cell"
    id="00"
  />
  <div
    class="cell"
    data-testid="cell"
    id="10"
  />
  <div
    class="cell"
    data-testid="cell"
    id="20"
  />
</div>
`;
